field/src/field.rs

```diff
@@ -151,24 +155,26 @@ pub trait PrimeField64: PrimeField {
         log2_ceil_u64(Self::ORDER_U64) as usize
     }
 
+    /// Return the representative of `value` that is less than `ORDER_U64`.
     fn as_canonical_u64(&self) -> u64;
+
+    /// Return the value \sum_{i=0}^N u[i] * v[i].
+    ///
+    /// NB: Assumes that sum(u) <= 2^32 or 2^64 depending on whether
+    /// the underlying field is 32-bit or 64-bit.
+    ///
+    /// TODO: Mark unsafe because of the assumption?
+    fn z_linear_combination_sml<const N: usize>(u: [u64; N], v: &[Self; N]) -> Self;
```

Comment:
Couple thoughts/questions on this,

I didn't get the meaning of `z` here?

It would seem more natural to me if `u` was `[Self; N]`, but I suppose the issue is that we want `u` to be encoded as canonical scalars, whereas `Self` might be Monty encoded?

The `32-bit or 64-bit` bit feels a bit odd, might be nice to split into two methods so they can declare more specific assumptions, although OTOH it would be more code. Might be worth considering separate traits for them, possibly with a blanket impl like `impl<F: SmallLc64> SmallLc32 for F`. But feel free to leave it as-is for now; I'll need to think more before having a strong opinion on the best API.

Reply:
The `z` just means "integers", i.e. I'm interpreting the elements of `u` as integers (canonical \ZZ-module structure yada yada) rather than from a particular field (or any particular representation thereof, such as Monty repr). This means I can store the MDS matrices as arrays `u64`s and use their values directly. Might be clearer if I just call the function `linear_combination_sml`.

Regarding the 32-bit vs 64-bit, I agree it feels a bit odd, and I only did it that way to express the most general conditions under which the function can be called. Given that the maximum value we need in `u` is 2^8 (or maybe 2^16), I will change `u`s type to `[u32; N]` which will enforce a suitable bound.